### DISCUSSION TITLE
attribute.type as function

### DISCUSSION THREAD

**Discussion by dandax** [markdoc/markdoc#54] [Q&A] [answered]
I'm looking at the source code and I noticed the Transform.ts attributes function has this:

```typescript
if (typeof attr.type === 'function') {
  const instance: any = new attr.type();
  if (instance.transform) {
    value = instance.transform(value, config);
  }
}
```

How would I use a function? I'm trying to create a custom `fence` schema and instead of the output of the `language` attribute being `data-language="{language}"` I want `class="language-{language}"`. Maybe I don't need a `transform` function for this so an easier solution would help, but also curious how I could use a function if I needed to in the future.

**Comment by mfix-stripe** [comment, marked as answer]
@dandax great question! You can just [custom attribute types](https://markdoc.io/docs/attributes#create-a-custom-attribute) to handle this. Here is a full example:

```js
class Language {
  transform(value) {
    return `language-${value}`;
  }
  validate(value) {
    // Add custom validation here!
  }
}

export default {
  render: 'Code',
  attributes: {
    content: { type: String },
    language: {
      render: 'class',
      type: Language,
      description:
        'The programming language of the code block. Place it after the backticks.'
    }
  }
};
```

Two things to note in this example:
1. `render` — the actual attribute name in your rendered output
2. Custom `Language` class, containing a `transform` function.
